File: core/src/net/brenig/pixelescape/game/player/abliity/AbilityBlink.java
package net.brenig.pixelescape.game.player.abliity;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.scenes.scene2d.utils.Drawable;

import net.brenig.pixelescape.PixelEscape;
import net.brenig.pixelescape.game.World;
import net.brenig.pixelescape.game.data.GameAssets;
import net.brenig.pixelescape.game.entity.particle.EntityFadingParticle;
import net.brenig.pixelescape.game.entity.EntityPlayer;
import net.brenig.pixelescape.lib.Reference;
import net.brenig.pixelescape.screen.GameScreen;

public class AbilityBlink extends Ability {

	private final static float maxCooldown = 10;
	private final static float range = Reference.BLOCK_WIDTH * 4;

	public AbilityBlink() {
		super(maxCooldown);
	}

	@Override
	public boolean onActivate(GameScreen screen, World world, EntityPlayer player) {
		for(int i = 0; i < 60; i++) {
			EntityFadingParticle e = world.createEntity(EntityFadingParticle.class);
			e.setPosition(player.getXPos() + PixelEscape.rand.nextFloat() * 20 - 10, player.getYPos() + PixelEscape.rand.nextFloat() * 40 - 20);
			e.setColor(Color.BLUE);
			e.setFadeDuration(0.4F);
			e.setAccelerationFactor(0.99F, 0.99F);
			e.setVelocity(PixelEscape.rand.nextFloat() * 10 - 0.5F, (PixelEscape.rand.nextFloat() * 20 + 20) * (e.getYPos() > player.getYPos() ? 1 : -1));
			world.spawnEntity(e);
		}
		player.increaseXPos(range);

		final float oldX = screen.worldRenderer.getTargetX();
		screen.worldRenderer.setCameraXPosition(screen.worldRenderer.getXPos() + range);
		screen.worldRenderer.moveScreenTo(oldX, Reference.BLOCK_WIDTH * 4);
		return true;
	}

	public Drawable getDrawable(GameAssets assets) {
		return assets.getItemBlink();
	}
}

File: core/src/net/brenig/pixelescape/game/player/abliity/AbilityBlink.kt
package net.brenig.pixelescape.game.player.abliity

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.utils.Drawable
import net.brenig.pixelescape.PixelEscape
import net.brenig.pixelescape.game.World
import net.brenig.pixelescape.game.data.GameAssets
import net.brenig.pixelescape.game.entity.impl.EntityPlayer
import net.brenig.pixelescape.game.entity.impl.particle.EntityFadingParticle
import net.brenig.pixelescape.game.data.constants.Reference
import net.brenig.pixelescape.screen.GameScreen

class AbilityBlink : Ability(maxCooldown) {

    override fun onActivate(screen: GameScreen, world: World, player: EntityPlayer): Boolean {
        for (i in 0..59) {
            val e = world.createEntity(EntityFadingParticle::class.java)
            e.setPosition(player.xPos + PixelEscape.rand.nextFloat() * 20 - 10, player.yPos + PixelEscape.rand.nextFloat() * 40 - 20)
            e.setColor(Color.BLUE)
            e.setFadeDuration(0.4f)
            e.setAccelerationFactor(0.99f, 0.99f)
            e.setVelocity(PixelEscape.rand.nextFloat() * 10 - 0.5f, (PixelEscape.rand.nextFloat() * 20 + 20) * if (e.yPos > player.yPos) 1 else -1)
            world.spawnEntity(e)
        }
        player.increaseXPos(range)

        val oldX = screen.worldRenderer.targetX
        screen.worldRenderer.setCameraXPosition(screen.worldRenderer.xPos + range)
        screen.worldRenderer.moveScreenTo(oldX, (Reference.BLOCK_WIDTH * 4).toFloat())
        return true
    }

    override fun getDrawable(assets: GameAssets): Drawable {
        return assets.itemBlink
    }

    companion object {

        private val maxCooldown = 10f
        private val range = (Reference.BLOCK_WIDTH * 4).toFloat()
    }
}
